### PR TITLE
feature/pwm-chip-by-name

### PR DESCRIPTION
**Pull request by otiv-jerome-colle**
Adds methods to instantiate chips by name, as they might get allocated dynamically

### CODE REVIEW

File: examples/breathe.rs [hunk 1]
```diff
@@ -19,25 +19,29 @@ const BB_PWM_NUMBER: u32 = 0;
 fn pwm_increase_to_max(pwm: &Pwm,
                        duration_ms: u32,
                        update_period_ms: u32) -> Result<()> {
-    let step: f32 = duration_ms as f32 / update_period_ms as f32;
+    let num_steps: f32 = duration_ms as f32 / update_period_ms as f32;
```
Comment (otiv-emiel-vanseveren) [inline examples/breathe.rs]: could have these both as durations, just double checked, durations have a safe division which can return an f32

File: examples/breathe.rs [hunk 2]
```diff
@@ -19,25 +19,29 @@ const BB_PWM_NUMBER: u32 = 0;
...
+    let step = 1.0 / num_steps;
     let mut duty_cycle = 0.0;
-    let period_ns: u32 = try!(pwm.get_period_ns());
+    let period_ns: u32 = pwm.get_period_ns()?;
     while duty_cycle < 1.0 {
-        try!(pwm.set_duty_cycle_ns((duty_cycle * period_ns as f32) as u32));
+        pwm.set_duty_cycle_ns((duty_cycle * period_ns as f32) as u32)?;
         duty_cycle += step;
+        std::thread::sleep(std::time::Duration::from_millis(update_period_ms as u64))
```
Comment (otiv-emiel-vanseveren) [inline examples/breathe.rs]: then this would also not be needed anymore

File: examples/breathe.rs [hunk 3]
```diff
@@ -19,25 +19,29 @@ const BB_PWM_NUMBER: u32 = 0;
 fn pwm_increase_to_max(pwm: &Pwm,
```
Comment (otiv-emiel-vanseveren) [inline examples/breathe.rs]: rename to maximum? to match the minimum function

File: examples/breathe.rs [hunk 4]
```diff
@@ -19,25 +19,29 @@ const BB_PWM_NUMBER: u32 = 0;
...
     while duty_cycle < 1.0 {
-        try!(pwm.set_duty_cycle_ns((duty_cycle * period_ns as f32) as u32));
+        pwm.set_duty_cycle_ns((duty_cycle * period_ns as f32) as u32)?;
         duty_cycle += step;
+        std::thread::sleep(std::time::Duration::from_millis(update_period_ms as u64))
     }
     pwm.set_duty_cycle_ns(period_ns)
 }
 
 fn pwm_decrease_to_minimum(pwm: &Pwm,
```
Comment (otiv-emiel-vanseveren) [inline examples/breathe.rs:36]: This function looks exactly the same as the duty_cycle, maybe refactor into one? not required ofc

File: src/error.rs [hunk 5]
```diff
@@ -22,14 +22,7 @@ pub enum Error {
 }
 
 impl ::std::error::Error for Error {
```
Comment (otiv-emiel-vanseveren) [inline src/error.rs]: never seen these '::' in front of std :p

File: src/lib.rs [hunk 6]
```diff
@@ -45,91 +45,137 @@ pub type Result<T> = ::std::result::Result<T, error::Error>;
...
     let mut s = String::with_capacity(10);
-    let mut f = try!(pwm_file_ro(chip, pin, name));
-    try!(f.read_to_string(&mut s));
+    let mut f = pwm_file_ro(chip, pin, name)?;
+    f.read_to_string(&mut s)?;
     match s.trim().parse::<T>() {
         Ok(r) => Ok(r),
-        Err(_) => Err(Error::Unexpected(format!("Unexpeted value file contents: {:?}", s))),
+        Err(_) => Err(Error::Unexpected(format!(
+            "Unexpeted value file contents: {:?}",
```
Comment (otiv-emiel-vanseveren) [inline src/lib.rs]: typo: unexpected

File: src/lib.rs [hunk 7]
```diff
@@ -45,91 +45,137 @@ pub type Result<T> = ::std::result::Result<T, error::Error>;
...
 impl PwmChip {
     pub fn new(number: u32) -> Result<PwmChip> {
-        try!(fs::metadata(&format!("/sys/class/pwm/pwmchip{}", number)));
-        Ok(PwmChip { number: number })
+        fs::metadata(format!("/sys/class/pwm/pwmchip{}", number))?;
+        Ok(PwmChip { number })
+    }
+
+    pub fn new_by_name(name: &str) -> Result<PwmChip> {
+        let pwmchip_number = Self::get_pwm_number_by_name(name).ok_or(Error::Unexpected(
```
Comment (otiv-emiel-vanseveren) [inline src/lib.rs:85]: these errors are veryy generic, maybe add an error type in the lines of Chip/pin name not found?

File: src/lib.rs [hunk 8]
```diff
@@ -45,91 +45,137 @@ pub type Result<T> = ::std::result::Result<T, error::Error>;
...
+        Ok(Self {
+            number: pwmchip_number,
+        })
     }
 
     pub fn count(&self) -> Result<u32> {
         let npwm_path = format!("/sys/class/pwm/pwmchip{}/npwm", self.number);
-        let mut npwm_file = try!(File::open(&npwm_path));
+        let mut npwm_file = File::open(&npwm_path)?;
         let mut s = String::new();
```
Comment (otiv-emiel-vanseveren) [inline src/lib.rs]: rename s to file_contents?

File: src/lib.rs [hunk 9]
```diff
@@ -45,91 +45,137 @@ pub type Result<T> = ::std::result::Result<T, error::Error>;
...
-        try!(npwm_file.read_to_string(&mut s));
+        npwm_file.read_to_string(&mut s)?;
         match s.parse::<u32>() {
             Ok(n) => Ok(n),
-            Err(_) => Err(Error::Unexpected(format!("Unexpected npwm contents: {:?}", s))),
+            Err(_) => Err(Error::Unexpected(format!(
+                "Unexpected npwm contents: {:?}",
```
Comment (otiv-emiel-vanseveren) [inline src/lib.rs:101]: similar here as previous comment, maybe have a specific error case for handling Unexecpted npwm content instead?

File: src/lib.rs [hunk 10]
```diff
@@ -45,91 +45,137 @@ pub type Result<T> = ::std::result::Result<T, error::Error>;
...
+                        .ok()?
+                        .strip_prefix("pwmchip")
+                        .and_then(|chip_str| chip_str.parse::<u32>().ok())
+                })
+            })
+    }
 }
 
 impl Pwm {
     /// Create a new Pwm wiht the provided chip/number
```
Comment (otiv-emiel-vanseveren) [inline src/lib.rs]: typo with

**Review by otiv-emiel-vanseveren — APPROVED** [review]

File: examples/breathe.rs (hunk 1 above)
Comment (otiv-jerome-colle) [reply examples/breathe.rs]: Cool, thanks for reviewing this! Changed in [6deebee](https://github.com/otiv-jerome-colle/rust-sysfs-pwm/pull/2/commits/6deebeea5622c9f0136742a3772b633e2a7b6a9d)

File: examples/breathe.rs (hunk 2 above)
Comment (otiv-jerome-colle) [reply examples/breathe.rs]: [6deebee](https://github.com/otiv-jerome-colle/rust-sysfs-pwm/pull/2/commits/6deebeea5622c9f0136742a3772b633e2a7b6a9d)

File: examples/breathe.rs (hunk 3 above)
Comment (otiv-jerome-colle) [reply examples/breathe.rs]: Yess [6deebee](https://github.com/otiv-jerome-colle/rust-sysfs-pwm/pull/2/commits/6deebeea5622c9f0136742a3772b633e2a7b6a9d)

File: examples/breathe.rs (hunk 4 above)
Comment (otiv-jerome-colle) [reply examples/breathe.rs:36]: This is just an example, for which I prefer it to be a bit bulkier but more readable. Two functions convey the meaning a bit faster imo

File: src/error.rs (hunk 5 above)
Comment (otiv-jerome-colle) [reply src/error.rs]: Old syntax, should've updated that originally

File: src/lib.rs (hunk 6 above)
Comment (otiv-jerome-colle) [reply src/lib.rs]: Fixed! [6deebee](https://github.com/otiv-jerome-colle/rust-sysfs-pwm/pull/2/commits/6deebeea5622c9f0136742a3772b633e2a7b6a9d)

File: src/lib.rs (hunk 7 above)
Comment (otiv-jerome-colle) [reply src/lib.rs:85]: Indeed, but I consider that out-of-scope for this PR. Feel free to make a new one to fix that. I'll do that later anyway

File: src/lib.rs (hunk 8 above)
Comment (otiv-jerome-colle) [reply src/lib.rs]: Done; [6deebee](https://github.com/otiv-jerome-colle/rust-sysfs-pwm/pull/2/commits/6deebeea5622c9f0136742a3772b633e2a7b6a9d)

File: src/lib.rs (hunk 9 above)
Comment (otiv-jerome-colle) [reply src/lib.rs:101]: I'll leave that to the other PR as well

File: src/lib.rs (hunk 10 above)
Comment (otiv-jerome-colle) [reply src/lib.rs]: Done! [6deebee](https://github.com/otiv-jerome-colle/rust-sysfs-pwm/pull/2/commits/6deebeea5622c9f0136742a3772b633e2a7b6a9d)

File: src/error.rs (hunk 5 above)
Comment (otiv-emiel-vanseveren) [reply src/error.rs]: can you update it? :p

**Review by otiv-emiel-vanseveren — APPROVED** [review]
just the ::std otherwise LGTM!

File: src/error.rs (hunk 5 above)
Comment (otiv-jerome-colle) [reply src/error.rs]: out-of-scope but I'll add it

**Pull request merged**